### PR TITLE
add vignette describing release process

### PR DESCRIPTION
**Pull request by zkamvar**
This adds a vignette to describe the release process based on https://github.com/orgs/hubverse-org/discussions/29.

### CODE REVIEW

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:54]: While the recap in semantic version is super useful, it might be better to use hubverse examples as points of reference

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:174]: Overall this schedule is not ideal for the hubverse as many hubs collect submissions on a wednesday so not the best day for releasing new versions of packages. Also given not all of us are full time/work every day of the week, I wonder if this is a bit too restrictive for us atm.

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:146]: I think many folks (including myself) use GitHub web interface for releases. Mine are signed because I have listed my GPG key on GitHub so might be nice to provide instructions for that too (as it would mean other online GitHub activities would also be signed). This could be a shorter separate article though.

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:23]: I'm a little confused on what happens if the feature I'm working on is spread across multiple PRs. Do I merge this into `release-*` branch? Also do I increment the dev version? (I would definitely prefer that so we can quickly pin point which version of code we're discussing at any point if needed)

**Review by annakrystalli — COMMENTED** [review]
I've made a first pass with some comments. Would be really interested in feedback from others too!

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:174]: I had a feeling that was the case; I definitely don't want to put too many restrictions on this so I'll go ahead and strike it out.

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:146]: Agreed to split this out to a shorter article.

I also realized that I misspoke. You are correct that your commits are signed on the GitHub interface, but your tags will not be. The releases themselves on GitHub are not signed and you need to explicitly create a signed tag on the CLI. Tags created via the GitHub web interface become [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags), which can easily be moved around (this is how a lot of GitHub actions specify their versions: they move the `v2` tag to new commits as they update).

I created an example of this in https://github.com/zkamvar/improved-enigma/releases and you can see whether or not a tag is signed if it has a green check mark next to it. 

### Release via GitHub interface: unsigned tag

release showing an unsigned tag on a signed commit

### Release via GitHub interface _after_ creating a signed tag:

I also created a release _after_ created a signed tag:

release showing a signed tag on a signed commit

### Moving a lightweight tag to a new unsigned commit after GitHub Release

And I created a release where I moved a lightweight tag from one signed commit to another unsigned commit:

a release with a lightweight tag on an unsigned commit with details showing that it was force-pushed

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:23]: > I'm a little confused on what happens if the feature I'm working on is spread across multiple PRs. 
> Do I merge this into `release-*` branch? 

This is a good question. The answer is no. The `release-X.Y.Z` branch is a temporary branch that has one job: to bump the version number. It's created when you want to make a release from `main` and then deleted after it's merged into main and then the GitHub release is created. 

As you work on multiple independent features, you would merge them into `main` and build up to the release. This way, the pull requests can easily indicate how the releases were created. 

See **for example** how the releases are created in The Carpentries: https://github.com/carpentries/sandpaper/releases/tag/0.16.5. In this release, there are seven pull requests that were merged into `main` between April 2024 and June 2024. On 2024-06-18, The `release-0.16.5` branch was created to do the final pull request before release, which revealed minor errors that were easily corrected. Once it was merged, the release was created. 

> Also do I increment the dev version? (I would definitely prefer that so we can quickly pin point which version of code we're discussing at any point if needed)

I've gone back and forth on this. I usually leave the dev version to X.Y.Z.9000 so that we know whether it's dev or release and have people report the output of `packageDescription(pkg)$GitHubSHA1`. With constantly incrementing the version number, there end up being near-constant merge conflicts (and it's often difficult to remember to actually do so).

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:54]: Oh good point! I think the best example might be the `schemas` repo since that has major version changes (albeit no minor version changes, but the line between minor and patch is often subjective)

Comment (zkamvar) [reply vignettes/relase-process.qmd:53]: Aaand I realized that I had forgotten to scrub mentions of `$OLDJOB` packages, which now puts your comment in even more urgent context 🙈

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd]: It has been struck out

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:33]: I would also appreciate a section here on how we manage testing interdependent packages, both the stable and development versions.

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:21]: What do you mean by `Remotes:` is up to date? Would it have been changed somewhere in the above steps described? If so where and why?

File: vignettes/relase-process.qmd
Comment (annakrystalli) [inline vignettes/relase-process.qmd:16]: I feel it would help to have 4 different and smaller sections with their own checklists that could help break down conceptually the general workflow.

1. First PR after a release: Describe the actions required, i.e. bump the version to a `devel` version, update the `NEWS.md` with a new `(development)` section etc.
2. Subsequent PRs. A few words about next PRs. Anything we need to be aware of?
3. Release PR: I think you've already covered this pretty well below
4.  Hotfixes on releases. Would be good to get this down on paper too so we know whet to do

**Review by annakrystalli — COMMENTED** [review]
Added a few more comments on where I still have some confusion/would appreciate spelling out in a bit more detail.

I was also wondering, would issue or PR templates help here? Atleast in the initial stages of getting used to the system?

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:16]: This structure would definitely make it more clear. I will implement it.

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:21]: This was a copy/paste artifact that describes an edge case (after development of cross-package functionality). 

Effectively, if you have two packages that have complementary features in two pull requests, you can test them in the pull request by changing the `Remotes:` field (this was before I fully understood the `Config/needs/*` section. 

```mermaid
flowchart LR
   user
   pkgA
   pkgB
   PRA1["pkgA@feature-A1"]
   PRB1["pkgB@feature-B1"]

   user-->|interface|pkgA
   pkgA-->|backend|pkgB
   PRB1-.->|pull request|pkgB
   PRA1-.->|pull request|pkgA
   PRA1-->|"Remotes:"|PRB1

```

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:23]: I think this guidance from the R packages book is relevant:

> Increment the development version, e.g. from 9000 to 9001, if you’ve added an important feature and you (or others) need to be able to detect or require the presence of this feature. For example, this can happen when two packages are developing in tandem. This is generally the only reason that we bother to increment the development version. This makes in-development versions special and, in some sense, degenerate. Since we don’t increment the development component with each Git commit, the same package version number is associated with many different states of the package source, in between releases.

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd:33]: That's a good point. I think I'm going to make a separate vignette. For now, I've created a repository that tests the releases for our universe weekly across windows, mac, and ubuntu (oldrel, release, and devel): https://github.com/zkamvar/universe-tester

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd]: I've removed it as bumping the dev version number is sufficient to account for this.

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd]: I've added a section at the bottom

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd]: I've added this guidance to the versioning section

File: vignettes/relase-process.qmd
Comment (zkamvar) [reply vignettes/relase-process.qmd]: I have prioritized mentioning the web interface for releases

File: .github/workflows/R-CMD-check.yaml
Comment (nickreich) [inline .github/workflows/R-CMD-check.yaml:1]: Just for my understanding here - these are changes to the template workflow that new packages would use (or existing packages if they update/import this workflow), and the cron job part ensures that even when the package isn't used for a while that it would occasionally be R CMD CHECKed for any problems?

File: DESCRIPTION
Comment (nickreich) [inline DESCRIPTION:4]: @zkamvar you should add yourself as an author here. And probably someone should be listed as a maintainer.

File: vignettes/release-process.qmd
Comment (nickreich) [inline vignettes/release-process.qmd:15]: For each of the vignettes, maybe we should say explicitly that what license(s) the original documents (and the new hubverse ones) were released under. Looks like a CC one from the original at least.

File: vignettes/release-checklists.qmd
Comment (nickreich) [inline vignettes/release-checklists.qmd:20]: Should individual non-tagged releases carry NEWS updates with them? Do we have guidance on whether these NEWS updates will be present already or whether the releaser will need to go through and write all of this up?

File: vignettes/release-checklists.qmd
Comment (nickreich) [inline vignettes/release-checklists.qmd:18]: maybe we could add a manual check to make sure that the author list has been updated based on contributions since the last version?

**Review by nickreich — CHANGES_REQUESTED** [review]
Overall, this looks great. I have just a few minor comments.

File: .github/workflows/R-CMD-check.yaml
Comment (zkamvar) [reply .github/workflows/R-CMD-check.yaml:1]: To answer your second question first: 
> the cron job part ensures that even when the package isn't used for a while that it would occasionally be R CMD CHECKed for any problems?

Yes. The CRAN ecosystem is always moving and a weekly check helps prevent breaking features from rearing their heads. 

> these 
are changes to the template workflow that new packages would use (or existing packages if they update/import this workflow),

No, these are the changes for the workflow for _this_ package as an example/confirmation. I will propose the changes across all existing packages. 

For future packages, the workflows that we currently import are all external, with the exception of the pkgdown preview action, which is from https://github.com/hubverse-org/hubverse-developer-actions  https://github.com/hubverse-org/hubDevs/blob/50aae00f17f0d90db9697e30a517b4d65da6fc49/R/use_hubdev_pkg_actions.R#L14-L16

That being said, we could copy over the workflows from r-lib/actions, but then we would have to make sure to maintain them in parallel. It might even be possible to make a GitHub workflow to copy the workflows from `r-lib/actions` and modify them with https://mikefarah.gitbook.io/yq, and save them into our `hubverse-developer-actions`.

File: vignettes/release-checklists.qmd
Comment (zkamvar) [reply vignettes/release-checklists.qmd:20]: Yes and yes in [lines 58 and 59](https://github.com/hubverse-org/hubDevs/pull/11/files/1ee04ba466e3ef849c9c5a1878ec6be5c4ac8662#diff-274d4788e65713582311e17f54c220db7e5bbdbf75d2d43afa425232520e8852R58-R59).

Also, I would avoid using the term "non-tagged releases" because if they aren't tagged, they aren't released. I would just call them pull requests.

File: vignettes/release-checklists.qmd
Comment (zkamvar) [reply vignettes/release-checklists.qmd:18]: That's a good suggestion! I'll make sure to add it!

File: vignettes/release-checklists.qmd
Comment (elray1) [inline vignettes/release-checklists.qmd:44]:
```suggestion
 - [ ] merge (with approval)
```

**Review by nickreich — APPROVED** [review]

**Review by LucieContamin — APPROVED** [review]
Thanks, looks good to me

**Comment by harryhoch** [comment]
A couple of comments:

1.  The description in the README is a bit vague. "The goal of hubDevs is to provide utilities for creating and standardising Hubverse packages."  Is this really all that this code is doing? It looks like it's setting up a new repository for hosting a new hub. This should be clarified.

2. The use of the GITHUB_PAT should be documented. Perhaps add a pointer to a page detailing how to configure the PAT?

3.  The output that I saw differed from what's in the README, reading as follows:

✔ Setting active project to "/Users/HARRYH/Downloads".
✔ Creating R/.
✔ Writing DESCRIPTION.
Package: Downloads
Title: What the Package Does (One Line, Title Case)
Version: 0.0.0.9000
Authors@R (parsed):
    * First Last <first.last@example.com> [aut, cre] (YOUR-ORCID-ID)
Description: What the package does (one paragraph).
License: `use_mit_license()`, `use_gpl3_license()` or friends to
    pick a license
Encoding: UTF-8
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.3.2
✔ Writing NAMESPACE.
✔ Writing Downloads.Rproj.
✔ Adding "^Downloads\\.Rproj$" to .Rbuildignore.
✔ Adding ".Rproj.user" to .gitignore.
✔ Adding "^\\.Rproj\\.user$" to .Rbuildignore.
✔ Setting active project to "<no active project>".
✔ Setting active project to "/Users/HARRYH/Downloads".
✔ Adding .DS_Store, .Rhistory, .Rdata, .httr-oauth, and .secrets to .gitignore and .Rbuildignore
✔ Adding testthat to Suggests field in DESCRIPTION.
✔ Adding "3" to Config/testthat/edition.
✔ Creating tests/testthat/.
✔ Writing tests/testthat.R.
☐ Call usethis::use_test() to initialize a basic test file and open it for editing.
✔ Adding "MIT + file LICENSE" to License.
✔ Writing LICENSE.
✔ Writing LICENSE.md.
✔ Adding "^LICENSE\\.md$" to .Rbuildignore.
✔ Writing README.Rmd.
✔ Adding "^README\\.Rmd$" to .Rbuildignore.
✔ Adding "Lifecycle: experimental badge" to README.Rmd.
☐ Re-knit README.Rmd with devtools::build_readme().
✔ Adding "CRAN status badge" to README.Rmd.
☐ Re-knit README.Rmd with devtools::build_readme().
ℹ Installing Downloads in temporary library
ℹ Building /Users/HARRYH/Downloads/README.Rmd
✔ Creating .github/.
✔ Adding "^\\.github$" to .Rbuildignore.
✔ Adding "*.html" to .github/.gitignore.
✔ Writing .github/CODE_OF_CONDUCT.md.
✔ Writing .github/CONTRIBUTING.md.
✔ Initialising Git repo.
✔ Adding ".quarto" to .gitignore.
ℹ There are 105 uncommitted files:
• .#notes
• .company.thebrowser.Browser.09iTRP
• .company.thebrowser.Browser.1mVc9x
• .company.thebrowser.Browser.1rraSJ
• .company.thebrowser.Browser.2gYBMK
• .company.thebrowser.Browser.2ObmWf
• .company.thebrowser.Browser.2SWvHY
• .company.thebrowser.Browser.2Za511
• .company.thebrowser.Browser.3OvsCV
• .company.thebrowser.Browser.42peGV
  … and 95 more
! Is it ok to commit them?

1: Yeah
2: Negative
3: No way

The Vignette might need more detail on what it does and how it works.

**Comment by zkamvar** [comment]
Thank you for the review of the README, @harryhoch. I'll go ahead and open an issue for it. This PR was focused on the release vignettes and since I've gotten passing reviews on this, I will merge it.

**Pull request merged**
